### PR TITLE
Add functional tests for richmedia module

### PR DESCRIPTION
**Pull request by osulzhenko**

### CODE REVIEW

File: src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy:288]: Just PATTERN_NAME

**Review by marki1an — CHANGES_REQUESTED** [review]

File: src/test/groovy/org/prebid/server/functional/model/response/auction/AppliedTo.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/model/response/auction/AppliedTo.groovy]: nice catch, updated

File: src/test/groovy/org/prebid/server/functional/model/request/auction/RichmediaFilter.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/model/request/auction/RichmediaFilter.groovy]: updated, also updated some test cases

File: src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy:28]: updated

File: src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy]: updated

File: src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy]: updated

File: pom.xml
Comment (osulzhenko) [reply pom.xml:117]: I removed the version for a particular dependencies to avoid conflicts with other dependencies. Specifically, there was a issue of versions collision between `pbs.jar` and `pbs-bundle.jar`.

File: src/test/groovy/org/prebid/server/functional/model/response/auction/InvocationResult.groovy
Comment (Net-burst) [inline src/test/groovy/org/prebid/server/functional/model/response/auction/InvocationResult.groovy:14]: I'm not sure that it's the same type. FetchStatus was made for Floors AFAIK. Module invocation status is a different thing. At least as far as I remember.

File: src/test/groovy/org/prebid/server/functional/model/response/auction/InvocationResult.groovy
Comment (Net-burst) [inline src/test/groovy/org/prebid/server/functional/model/response/auction/InvocationResult.groovy:16]: Variable name.

File: src/test/groovy/org/prebid/server/functional/model/response/auction/TraceModule.groovy
Comment (Net-burst) [inline src/test/groovy/org/prebid/server/functional/model/response/auction/TraceModule.groovy:9]: `ModuleTrace` looks more logical.

**Review by Net-burst — CHANGES_REQUESTED** [review]
Looks very good overall. Mostly minor remarks.

File: src/test/groovy/org/prebid/server/functional/model/response/auction/TraceModule.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/model/response/auction/TraceModule.groovy]: updated

File: src/test/groovy/org/prebid/server/functional/model/response/auction/InvocationResult.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/model/response/auction/InvocationResult.groovy]: updated

File: src/test/groovy/org/prebid/server/functional/model/response/auction/InvocationResult.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/model/response/auction/InvocationResult.groovy]: good catch, changed to a new object `InvocationStatus`

File: src/test/groovy/org/prebid/server/functional/model/response/auction/Bid.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/model/response/auction/Bid.groovy:20]: [The richModule filtering requires a pattern string in the bid request’s adm.](https://github.com/prebid/prebid-server/issues/2483#issue-1480186686):
`substring search the creative body (adm) for the pattern mraid.js. If it doesn't exist, go on the next bid response.`
We are also using `String adm;` for PBS’s `com.iab.openrtb.response.Bid.java`. Therefore, I have changed it to a String. I have checked the tests that required an `Adm` object, and they are functioning without any issues.

File: src/test/groovy/org/prebid/server/functional/model/config/AccountHooksConfiguration.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/model/config/AccountHooksConfiguration.groovy]: that would be great, but we already have `AccountGppConfig.ModuleConfig` for `activityConfig` =( Changed to `PrebidModulesConfig` is this okay?

File: src/test/groovy/org/prebid/server/functional/model/config/ExecutionGroup.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/model/config/ExecutionGroup.groovy]: updated

File: src/test/groovy/org/prebid/server/functional/model/response/auction/AppliedTo.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/model/response/auction/AppliedTo.groovy:13]: we discussed it with @AntoxaAntoxic, it should be a List. So, I returned it back and discussed with @marki1an

File: pom.xml
Comment (osulzhenko) [reply pom.xml:615]: Fortunately, it is not an issue. We compile the tests not with the `maven-surefire-plugin` or `maven-failsafe-plugin`, but with the `maven-compiler-plugin`, which hasn’t changed. I tested it with the fully disabled `failsafe` plugin. The only way we can skip compiling tests is by default using `-Dmaven.test.skip=true` in commands. Our custom variables `skipFunctionalTests` and `skipModuleFunctionalTests` can’t cause skip test compilation issue.

File: src/test/groovy/org/prebid/server/functional/model/config/AccountHooksConfiguration.groovy
Comment (Net-burst) [reply src/test/groovy/org/prebid/server/functional/model/config/AccountHooksConfiguration.groovy]: The most correct course of action would be to rename the `AccountGppConfig.ModuleConfig` to `AccountGppConfig.GppModuleConfig`. As for `AccountHooksConfiguration. PrebidModulesConfig` - it doesn't make much sense, as it's a config for PBS module configuration. There is no such thing as a Prebid module.
I propose this one of those:
1) `AccountGppConfig.ModuleConfig` -> `AccountGppConfig.GppModuleConfig` and `AccountHooksConfiguration.HooksModulesConfig` -> `AccountHooksConfiguration.ModulesConfig`
2) `AccountGppConfig.ModuleConfig` -> `AccountGppConfig.GppModuleConfig` and `AccountHooksConfiguration.HooksModulesConfig` -> `AccountHooksConfiguration.PbsModulesConfig` or `AccountHooksConfiguration.ServerModulesConfig`

File: src/test/groovy/org/prebid/server/functional/model/config/Stage.groovy
Comment (Net-burst) [reply src/test/groovy/org/prebid/server/functional/model/config/Stage.groovy:16]: This is still not done.

File: src/test/groovy/org/prebid/server/functional/model/config/AccountHooksConfiguration.groovy
Comment (Net-burst) [reply src/test/groovy/org/prebid/server/functional/model/config/AccountHooksConfiguration.groovy]: Another thing is that you are mixing the model for account config and request. Shouldn't they be completely separate?

File: src/test/groovy/org/prebid/server/functional/model/request/auction/Prebid.groovy
Comment (Net-burst) [inline src/test/groovy/org/prebid/server/functional/model/request/auction/Prebid.groovy:39]: I'll leave a separate remark here. Looks like you're reusing a class between request and DB config. This looks incorrect AFAIK. I don't think that all modules that could have account config will be able to take the *same* in the request.

File: src/test/groovy/org/prebid/server/functional/model/config/Stage.groovy
Comment (Net-burst) [inline src/test/groovy/org/prebid/server/functional/model/config/Stage.groovy:16]: I'll leave a separate comment again. Please arrange those in the order occur in the PBS.

File: src/test/groovy/org/prebid/server/functional/tests/module/ModuleBaseSpec.groovy
Comment (Net-burst) [inline src/test/groovy/org/prebid/server/functional/tests/module/ModuleBaseSpec.groovy:12]: Please remove the comment.

**Review by Net-burst — CHANGES_REQUESTED** [review]

File: src/test/groovy/org/prebid/server/functional/tests/module/ModuleBaseSpec.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/tests/module/ModuleBaseSpec.groovy]: removed

File: src/test/groovy/org/prebid/server/functional/model/config/Stage.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/model/config/Stage.groovy]: I didn’t see that comment, updated.

Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/model/config/Stage.groovy]: updated

File: src/test/groovy/org/prebid/server/functional/model/request/auction/Prebid.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/model/request/auction/Prebid.groovy:39]: updated

File: docs/developers/functional-tests.md
Comment (osulzhenko) [reply docs/developers/functional-tests.md]: updated

File: docs/developers/functional-tests.md
Comment (osulzhenko) [reply docs/developers/functional-tests.md]: updated

File: docs/developers/functional-tests.md
Comment (osulzhenko) [reply docs/developers/functional-tests.md]: updated

File: docs/developers/functional-tests.md
Comment (osulzhenko) [reply docs/developers/functional-tests.md]: updated

File: docs/developers/functional-tests.md
Comment (osulzhenko) [reply docs/developers/functional-tests.md]: updated

File: docs/developers/functional-tests.md
Comment (osulzhenko) [reply docs/developers/functional-tests.md]: updated

File: src/test/groovy/org/prebid/server/functional/model/request/auction/Prebid.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/model/request/auction/Prebid.groovy]: updated

File: src/test/groovy/org/prebid/server/functional/model/config/AccountHooksConfiguration.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/model/config/AccountHooksConfiguration.groovy]: updated

File: Dockerfile-module
Comment (osulzhenko) [reply Dockerfile-module]: updated

**Comment by osulzhenko** [comment]
> Would be greate to have a test when `hooks.pb-richmedia-filter.enabled : false`. I don't see test when in account and in host config feature enabled and from where should we get precedence?

but we already had that
host true, account false: 
`PBS should process request without analytics when adm matches with pattern name and filter enabled in host config but disabled in account config`
host false, account true:
`PBS should reject request with error and provide analytic when adm matches with pattern name and filter disabled in host config but enabled in account config`
Extended with host true, account true

File: src/test/groovy/org/prebid/server/functional/model/response/auction/Bid.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/model/response/auction/Bid.groovy:66]: updated

File: src/test/groovy/org/prebid/server/functional/model/response/auction/Bid.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/model/response/auction/Bid.groovy:71]: updated

File: src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy]: updated

File: src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy]: updated

File: src/test/groovy/org/prebid/server/functional/tests/module/ModuleBaseSpec.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/tests/module/ModuleBaseSpec.groovy]: updated

File: src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy]: updated

**Comment by marki1an** [comment]
> > Would be greate to have a test when `hooks.pb-richmedia-filter.enabled : false`. I don't see test when in account and in host config feature enabled and from where should we get precedence?
> 
> but we already had that host true, account false: `PBS should process request without analytics when adm matches with pattern name and filter enabled in host config but disabled in account config` host false, account true: `PBS should reject request with error and provide analytic when adm matches with pattern name and filter disabled in host config but enabled in account config` Extended with host true, account true

I'm not mean "hooks.**modules**.pb-richmedia-filter.filter-mraid": true/false, 
I offer this one property hooks.pb-richmedia-filter.enabled : true/false` entire hook.

**Comment by osulzhenko** [comment]
> > > Would be greate to have a test when `hooks.pb-richmedia-filter.enabled : false`. I don't see test when in account and in host config feature enabled and from where should we get precedence?
> > 
> > 
> > but we already had that host true, account false: `PBS should process request without analytics when adm matches with pattern name and filter enabled in host config but disabled in account config` host false, account true: `PBS should reject request with error and provide analytic when adm matches with pattern name and filter disabled in host config but enabled in account config` Extended with host true, account true
> 
> I'm not mean `hooks.module.pb-richmedia-filter.filter-mraid: true/false`, I offer this one property `hooks.pb-richmedia-filter.enabled : true/false` entire hook.

added

File: src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy:261]: At this test you can't know who will be prioritize when host config disabled

File: src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy:221]: Same here

**Review by marki1an — CHANGES_REQUESTED** [review]

File: src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy:221]: updated

File: src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy:261]: updated

**Review by Net-burst — APPROVED** [review]
LGTM after @marki1an also gives his approve.

**Review by SerhiiNahornyi — CHANGES_REQUESTED** [review]

File: src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy
Comment (SerhiiNahornyi) [inline src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy:21]: @AntoxaAntoxic Pls review this file

File: src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy
Comment (AntoxaAntoxic) [inline src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy:104]: imho looks too specific to be covered by functional tests

File: src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy
Comment (AntoxaAntoxic) [reply src/test/groovy/org/prebid/server/functional/tests/module/richmedia/RichMediaFilterSpec.groovy:261]: imho test name is a bit vague for me, maybe makes sense to mention that the pattern matches with the host pattern while both configs are enabled

**Review by AntoxaAntoxic — APPROVED** [review]

**Pull request merged**
